### PR TITLE
Implement message buffering to make sending non-blocking

### PR DESCRIPTION
**Pull request by elsirion**
Previously the network stack could block when trying to send too many messages to a TCP connection that could not send them out quickly enough. In theory that could lead to deadlocks, in practice it was hidden by the channels sending messages between main and io threads being large.

This PR makes sending messages non-blocking by buffering them inside the io thread. To avoid DoS two two limits are introduced:

1. Maximum number of unsent messages: if a connection can't send messages quickly enough and messages pile up it gets closed and re-established, hopefully this helps.
2. Maximum queued epochs: while there is no connection we still queue messages in the hope it gets re-established quickly. If there are messages from too many epochs in the buffer we drop it and won't buffer anything till there is a new connection. 

TODO:
- [x] Implement message buffer
- [x] Drop connections with overfull send buffers
- [x] Drop sessions with overfull resend buffers
- [x] Keep them dropped till reconnect
- [ ] Tests

### CODE REVIEW

File: fedimint-server/src/net/peers.rs [hunk 1]
```diff
@@ -85,6 +87,9 @@ struct PeerConnectionStateMachine<M> {
 
 struct CommonPeerConnectionState<M> {
     resend_queue: MessageQueue<M>,
```
Comment (dpc) [inline fedimint-server/src/net/peers.rs:89]: Should this be named "outgoing_queue"? It's not like it only contains items already sent/needing resent.

**Review by dpc — COMMENTED** [review]
I think I understand things here and it looks reasonable.

File: fedimint-server/src/net/peers.rs (hunk 1 above)
Comment (elsirion) [reply fedimint-server/src/net/peers.rs:89]: Good catch! It got multiple functions now, both keeping un-ACKed messages and to-be-sent messages.

**Review by dpc — DISMISSED** [review]

File: fedimint-server/src/net/queue.rs [hunk 2]
```diff
@@ -60,49 +84,209 @@ where
...
+    /// messages in each epoch and the epoch only increases.
+    pub fn buffered_epochs(&self) -> u64 {
+        // We only have non-epoch messages during DKG and on re-joins. Both happen infrequently
+        // enough that this O(n) search for the lowest and highest epoch number in the buffer is
+        // sufficient
+        let maybe_oldest = self
+            .queue
+            .iter()
+            .filter_map(|msg| msg.msg.message_epoch())
+            .next();
```
Comment (Maan2003) [inline fedimint-server/src/net/queue.rs:156]: is clippy working in CI?
it should be complaining about filter_map.next => find_map

**Review by Maan2003 — APPROVED** [review]
LGTM

File: fedimint-server/src/net/queue.rs (hunk 2 above)
Comment (elsirion) [reply fedimint-server/src/net/queue.rs]: fixed

**Comment by justinmoon** [comment]
I don't have enough context to review this

**Comment by elsirion** [comment]
Superseded by #1740

**Pull request closed without merging**